### PR TITLE
Multiple client connection attempts if hostname resolves to multiple addresses

### LINKED ISSUE #6698: Clickhouse client uses only first resolved ip from localhost
```
# clickhouse-client -h localhost
Code: 210. DB::NetException: Connection refused (localhost:9000)

# clickhouse-client -h ::1
Code: 210. DB::NetException: Connection refused (::1:9000)

# clickhouse-client -h 127.0.0.1
Connected to ClickHouse server version 19.11.3 revision 54423.

# getent ahosts localhost 
::1             STREAM localhost
::1             DGRAM  
::1             RAW    
127.0.0.1       STREAM 
127.0.0.1       DGRAM  
127.0.0.1       RAW    

# getent hosts localhost 
::1             localhost localhost.localdomain localhost6 localhost6.localdoman6


# ping localhost
PING localhost (127.0.0.1) 56(84) bytes of data.
64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=0.024 ms
```

Probably same issue is possible for other connection types (for example intercluster server 2  server).

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Client will try every IP address returned by DNS resolution until successful connection.

closes #6698

### CODE REVIEW

File: src/Client/Connection.cpp [hunk 1]
```diff
@@ -93,37 +93,59 @@ void Connection::connect(const ConnectionTimeouts & timeouts)
...
+                    throw;
+                continue;
+            }
+            catch (Poco::TimeoutException &)
+            {
+                if (++it == addresses.end())
+                    throw;
+                continue;
+            }
+            break;
```
Comment (antonio2368) [inline src/Client/Connection.cpp:146]: maybe put break inside the `try` to make it more clear?

File: src/Client/Connection.cpp [hunk 2]
```diff
@@ -93,37 +93,59 @@ void Connection::connect(const ConnectionTimeouts & timeouts)
...
-        if (static_cast<bool>(secure))
+        auto addresses = DNSResolver::instance().resolveAddressList(host, port);
+        const auto & connection_timeout = static_cast<bool>(secure) ? timeouts.secure_connection_timeout : timeouts.connection_timeout;
+
+        for (auto it = addresses.begin(); it != addresses.end();)
         {
+            if (connected)
+                disconnect();
+
+            current_resolved_address = *it;
```
Comment (antonio2368) [inline src/Client/Connection.cpp:110]: We can assign after successfully connecting with the address

File: src/Common/DNSResolver.cpp [hunk 3]
```diff
@@ -208,6 +215,25 @@ Poco::Net::SocketAddress DNSResolver::resolveAddress(const std::string & host, U
...
+        return std::vector<Poco::Net::SocketAddress>{{ip, port}};
+
+    std::vector<Poco::Net::SocketAddress> addresses;
+
+    if (!impl->disable_cache)
+        addToNewHosts(host);
+
+    std::vector<Poco::Net::IPAddress> ips = impl->disable_cache ? hostByName(host) : impl->cache_host(host);
+    auto ips_end = std::unique(ips.begin(), ips.end());
+
```
Comment (antonio2368) [inline src/Common/DNSResolver.cpp:230]:
```suggestion
    addresses.reserve(ips_end - ips.begin());
```

File: src/Client/Connection.cpp [hunk 4]
```diff
@@ -93,37 +93,58 @@ void Connection::connect(const ConnectionTimeouts & timeouts)
...
             static_cast<bool>(compression) ? "" : ". Uncompressed");
 
-        if (static_cast<bool>(secure))
+        auto addresses = DNSResolver::instance().resolveAddressList(host, port);
+        const auto & connection_timeout = static_cast<bool>(secure) ? timeouts.secure_connection_timeout : timeouts.connection_timeout;
+
+        for (auto it = addresses.begin(); it != addresses.end();)
         {
+            if (connected)
+                disconnect();
```
[the remark this answers is not in the document]
Comment (yakov-olkhovskiy) [reply src/Client/Connection.cpp:108]: I believe after every connection attempt it's better to reset connection for good measure

**Review by antonio2368 — APPROVED** [review]

**Comment by yakov-olkhovskiy** [comment]
@mergify update

**Pull request merged**